roll/websockets.py

```diff
@@ -0,0 +1,147 @@
+import asyncio
+import types
+
+from websockets import handshake, WebSocketCommonProtocol, InvalidHandshake
+from websockets import ConnectionClosed  # exposed for convenience
+
+from roll import ProtocolUpgrade
+
+
+def create_websocket(
+        transport, subprotocol,
+        websocket_timeout=5,
+        websocket_max_size=2 ** 20,  # 1 megabytes
+        websocket_max_queue=16,
+        websocket_read_limit=2 ** 16,
+        websocket_write_limit=2 ** 16):
+    """Instanciate a new websocket with the given subprotocol.
+    """
+    websocket = WebSocketCommonProtocol(
+        timeout=websocket_timeout,
+        max_size=websocket_max_size,
+        max_queue=websocket_max_queue,
+        read_limit=websocket_read_limit,
+        write_limit=websocket_write_limit
+    )
+    websocket.subprotocol = subprotocol
+    websocket.connection_made(transport)
+    websocket.connection_open()
+    return websocket
+
+
+class WebsocketProtocol(ProtocolUpgrade):
+
+    __slots__ = ('websocket')
+
+    def __init__(self, request, subprotocols, **params):
+        self.subprotocol = None
+        if subprotocols:
+            ws_protocol = request.headers.get('SEC-WEBSOCKET-PROTOCOL')
+            if ws_protocol:
+                # select a subprotocol
+                client_subprotocols = tuple(
+                    (p.strip() for p in ws_protocol.split(',')))
+                for p in client_subprotocols:
+                    if p in subprotocols:
+                        self.subprotocol = p
+                        break
+
+    def connection_lost(self, protocol, exc):
+        self.websocket.connection_lost(exc)
+    connection_lost.bubble_up = True
+
+    def data_received(self, protocol, data):
+        # Received data. We refuse the data if the websocket is
+        # already closed. If the websocket is closing, this data
+        # might be part of the closing handshake (closing frame)
+        if self.websocket.state != 3:  # not closed
```

Comment:
Can we set that magic number as a constant somewhere?

Reply:
My apologies, it exists in the "websockets" package
https://github.com/aaugustin/websockets/blob/master/websockets/protocol.py#L33